lib/download/downloads_cache.rb

```diff
@@ -1,35 +1,74 @@
-    @dsd = nil
-    if path.nil?
-      @got_handle ||= {}
-      @dsd = @got_handle[handle] || Download.get(handle)
-      raise "handle '#{handle}' does not exist" if @dsd.nil?
-      path = @dsd.extract_path_flex.blank? ? @dsd.save_path_flex : @dsd.extract_path_flex
-      @got_handle[handle] = @dsd
+  def initialize(options = nil)
+    @cache = { dloads: {}, dsds: {}, used_dloads: {} }
+    @dload = @data_source = ds_id = nil
+    @options = options
+    if options
+      ds_id = options.delete(:data_source)  ## get DS id (if any) and also remove from options hash
+      @options = Hash[options.sort].to_json.downcase  ## slick. serialize hash in key-sorted order. -dji
     end
-    
-    @cache_handle = path
-    @handle = handle
-    @sheet = (@dsd.nil? || @dsd.sheet_override.blank?) ? sheet : @dsd.sheet_override.strip
-    file_key = make_cache_key('xls', @cache_handle)
-    sheet_key = make_cache_key('xls', @cache_handle, @sheet)
-
-    #if handle in cache, it was downloaded recently... need to pull this handle logic out to make less hacky
-    if handle != 'manual' && !Rails.cache.exist?(file_key)
-      Rails.logger.debug "!!! xls cache miss for file_key=#{file_key}"
-      download_handle
-      set_files_cache(file_key, 1)  ## Marker to show that file is downloaded
+    if ds_id
+      @data_source = DataSource.find(ds_id) || raise("No data source with id='#{ds_id}' found")
     end
-    get_files_cache(sheet_key) || set_xls_sheet(@sheet, date)
+  end
+
+  def setup_and_check(type, handle, path = nil, skip_proc = false)
+    Rails.logger.debug { "... Entered method setup_and_check: type=#{type}, handle=#{handle}, path=#{path}" }
+    skip = false
+    if path.nil?  ## this means that handle != 'manual'
+      @dload = @cache[:dloads][handle] || Download.get(handle) || raise("handle '#{handle}' does not exist")
+      @cache[:dloads][handle] = @dload
+      @handle = handle
+      path = @dload.extract_path_flex.blank? ? @dload.save_path_flex : @dload.extract_path_flex
+      cache_key = make_cache_key(type, path)
+      unless files_cache_exists? cache_key
+        download_handle
+        set_files_cache(cache_key, 1) if type == 'xls' ## Marker to show that xls file is downloaded
+      end
+      ## Now, figure out if we can skip over this source entirely because it hasn't changed.
+      if @data_source && skip_proc
+        bridge_key = @data_source.id.to_s + '_' + @dload.id.to_s
+        dsd = @cache[:dsds][bridge_key] || DataSourceDownload.get_or_new(@data_source.id, @dload.id)
+        @cache[:dsds][bridge_key] = dsd
+
+        skip = @dload.last_change_at <= dsd.last_file_vers_used && @options == dsd.last_eval_options_used
+      end
+    end
+    @path = path
+    skip
+  end
+
+  def mark_handle_used(handle)
+    @cache[:used_dloads][handle] = @cache[:dloads][handle] || raise("No download handle #{handle} to mark")
+  end
+
+  def update_last_used
+    return if @cache[:dsds].empty?
```

Comment:
`@cache[:dsds].empty?` is true after I updated the eval for DataSource 206052. As a result the skip check at line 35 is always false after making a change to the eval.